AI_LAB_PROJECT
Members: Muhammad Haris Hamid (24F-0699)
         Muhammad Ahmad (24F-0513)
         Waleed Bin Nasir (24F-0516)

PHASE 1
Python and Data Foundation

Step 1: Installing and importing libraries

In [3]:
#importing pandas library
import pandas as pd

Step 2: Loading Dataset

In [4]:
#loading our data set (Track D)
df=pd.read_csv("mental-heath-in-tech-2016_20161114.csv")

Step 3: Inspecting Dataset 

In [5]:
#printing the first 5 rows using head 
df.head()

,Are you self-employed?,How many employees does your company or organization have?,Is your employer primarily a tech company/organization?,Is your primary role within your company related to tech/IT?,Does your employer provide mental health benefits as part of healthcare coverage?,Do you know the options for mental health care available under your employer-provided coverage?,"Has your employer ever formally discussed mental health (for example, as part of a wellness campaign or other official communication)?",Does your employer offer resources to learn more about mental health concerns and options for seeking help?,Is your anonymity protected if you choose to take advantage of mental health or substance abuse treatment resources provided by your employer?,"If a mental health issue prompted you to request a medical leave from work, asking for that leave would be:",...,"If you have a mental health issue, do you feel that it interferes with your work when being treated effectively?","If you have a mental health issue, do you feel that it interferes with your work when NOT being treated effectively?",What is your age?,What is your gender?,What country do you live in?,What US state or territory do you live in?,What country do you work in?,What US state or territory do you work in?,Which of the following best describes your work position?,Do you work remotely?
0,0,26-100,1.0,NaN,Not eligible for coverage / N/A,NaN,No,No,I don't know,Very easy,...,Not applicable to me,Not applicable to me,39,Male,United Kingdom,NaN,United Kingdom,NaN,Back-end Developer,Sometimes
1,0,6-25,1.0,NaN,No,Yes,Yes,Yes,Yes,Somewhat easy,...,Rarely,Sometimes,29,male,United States of America,Illinois,United States of America,Illinois,Back-end Developer|Front-end Developer,Never
2,0,6-25,1.0,NaN,No,NaN,No,No,I don't know,Neither easy nor difficult,...,Not applicable to me,Not applicable to me,38,Male,United Kingdom,NaN,United Kingdom,NaN,Back-end Developer,Always
3,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,Sometimes,Sometimes,43,male,United Kingdom,NaN,United Kingdom,NaN,Supervisor/Team Lead,Sometimes
4,0,6-25,0.0,1.0,Yes,Yes,No,No,No,Neither easy nor difficult,...,Sometimes,Sometimes,43,Female,United States of America,Illinois,United States of America,Illinois,Executive Leadership|Supervisor/Team Lead|Dev ...,Sometimes


In [6]:
#printing the shape of the data set 
print(df.shape)

(1433, 63)


In [7]:
#checking missing values in the data set
df.isnull().sum()

Are you self-employed?                                                                  0
How many employees does your company or organization have?                            287
Is your employer primarily a tech company/organization?                               287
Is your primary role within your company related to tech/IT?                         1170
Does your employer provide mental health benefits as part of healthcare coverage?     287
                                                                                     ... 
What US state or territory do you live in?                                            593
What country do you work in?                                                            0
What US state or territory do you work in?                                            582
Which of the following best describes your work position?                               0
Do you work remotely?                                                                   0
Length: 63

In [8]:
#printing Unique values of target column
target_col="Have you been diagnosed with a mental health condition by a medical professional?"
print(f"Our Target column is: {target_col}")
print("Unique values in our target column are as follows :")
print(df[target_col].unique())

Our Target column is: Have you been diagnosed with a mental health condition by a medical professional?
Unique values in our target column are as follows :
<StringArray>
['Yes', 'No']
Length: 2, dtype: str


Step 4: Apply Python Fundamentals

In [9]:
#target column's different classes(response) and their count
#we define an empty dictionary to store the result 
target_col_class_count={}
for value in df[target_col]:
    #if there are any null values or blanks we convert them to string to make sure that it doesnt crash
    value=str(value)
    if value in target_col_class_count:
        target_col_class_count[value]+=1
    else:
        target_col_class_count[value]=1
print("The class Distribution in target col is as follows:")
print(target_col_class_count)        
        

The class Distribution in target col is as follows:
{'Yes': 716, 'No': 717}


In [10]:
# we store the columns of our dataset into a list
columns_list=list(df.columns)
print(columns_list)

['Are you self-employed?', 'How many employees does your company or organization have?', 'Is your employer primarily a tech company/organization?', 'Is your primary role within your company related to tech/IT?', 'Does your employer provide mental health benefits as part of healthcare coverage?', 'Do you know the options for mental health care available under your employer-provided coverage?', 'Has your employer ever formally discussed mental health (for example, as part of a wellness campaign or other official communication)?', 'Does your employer offer resources to learn more about mental health concerns and options for seeking help?', 'Is your anonymity protected if you choose to take advantage of mental health or substance abuse treatment resources provided by your employer?', 'If a mental health issue prompted you to request a medical leave from work, asking for that leave would be:', 'Do you think that discussing a mental health disorder with your employer would have negative cons

In [11]:
#slicing the first 50 columns 
sample_data=df[:50]
print(sample_data)

    Are you self-employed?  \
0                        0   
1                        0   
2                        0   
3                        1   
4                        0   
5                        0   
6                        0   
7                        0   
8                        0   
9                        1   
10                       0   
11                       0   
12                       0   
13                       0   
14                       0   
15                       0   
16                       0   
17                       0   
18                       1   
19                       0   
20                       0   
21                       0   
22                       0   
23                       0   
24                       1   
25                       0   
26                       0   
27                       0   
28                       0   
29                       0   
30                       0   
31                       0   
32        

In [12]:
# function to describe the data set 
def describe_dataset():
    print("Dataset Description\n")
    print(f"Dataset Shape: {df.shape}\n")
    print(f"Dataset Column Names: {columns_list}\n")
    print(f"Dataset Class Distribution: {target_col_class_count}\n")

#calling the function
describe_dataset()
    

Dataset Description

Dataset Shape: (1433, 63)

Dataset Column Names: ['Are you self-employed?', 'How many employees does your company or organization have?', 'Is your employer primarily a tech company/organization?', 'Is your primary role within your company related to tech/IT?', 'Does your employer provide mental health benefits as part of healthcare coverage?', 'Do you know the options for mental health care available under your employer-provided coverage?', 'Has your employer ever formally discussed mental health (for example, as part of a wellness campaign or other official communication)?', 'Does your employer offer resources to learn more about mental health concerns and options for seeking help?', 'Is your anonymity protected if you choose to take advantage of mental health or substance abuse treatment resources provided by your employer?', 'If a mental health issue prompted you to request a medical leave from work, asking for that leave would be:', 'Do you think that discussin

Step 5: Object-Oriented Representation

In [13]:
# we define our class DataRecord
class DataRecord:
    def __init__(self,record_id,features,label):
        self.record_id=record_id
        self.features=features
        self.label=label
    def Display(self):
        print(f"Record ID: {self.record_id}\n")
        print(f"Features: {self.features}\n")
        print(f"Label: {self.label}\n")
        print("-"*15)
#we make an empty python list to store our objects 
object_list=[]
#for loop to iterate over 5 rows
for row_index in range(5):
    #we use iloc() to grab the row with index row_index which we just defined in our for loop (eg= row0,row1,..etc) inside our row variable
    row=df.iloc[row_index]
    #we store the index of our row as record id in a new variable 
    obj_id=row_index
    #we store the label as our target column since that column is our main focus 
    obj_label=row[target_col]
    #we store rest of the columns as a separate feature in the form of a dictionary to record feature and the response of the object
    #we also drop target_col since we already stored it as our label 
    obj_features=row.drop(target_col).to_dict()
    #we pass the values into the constructor of our object 
    new_record_object=DataRecord(record_id=obj_id,label=obj_label,features=obj_features)
    #we add that item into our list of objects (object_list)
    object_list.append(new_record_object)
#we iterate over our object list and call display for each object
for obj in object_list:
    obj.Display()


Record ID: 0

Features: {'Are you self-employed?': 0, 'How many employees does your company or organization have?': '26-100', 'Is your employer primarily a tech company/organization?': 1.0, 'Is your primary role within your company related to tech/IT?': nan, 'Does your employer provide mental health benefits as part of healthcare coverage?': 'Not eligible for coverage / N/A', 'Do you know the options for mental health care available under your employer-provided coverage?': nan, 'Has your employer ever formally discussed mental health (for example, as part of a wellness campaign or other official communication)?': 'No', 'Does your employer offer resources to learn more about mental health concerns and options for seeking help?': 'No', 'Is your anonymity protected if you choose to take advantage of mental health or substance abuse treatment resources provided by your employer?': "I don't know", 'If a mental health issue prompted you to request a medical leave from work, asking for that l

Step 6: Build a Graph from Your Data

In [37]:
#we need to build a graph between 2 cols we've used are given below in col1 and col2
#the relation ship the above columns represent is how soughting help for mental health affects work if you have 
col1="Have you ever sought treatment for a mental health issue from a mental health professional?"
col2="If you have a mental health issue, do you feel that it interferes with your work when NOT being treated effectively?"
#we define an empty dictionary for our graph the dictionary's key will be the unique values of each column and the values will be 
#a list which will contain all the relationships col1 has with col2 i.e neighbors eg= col1=Yes and col2=No is one relationship 
#col1=Yes and col2=Rarely is another one so node col1 has neighbors No and Rarely 
graph={}
for index in range(len(df)):
  #we grab the current row in our dataset
    row=df.iloc[index]
 #since col1 also has 1 or 0 as vals we convert it into str to check it
    raw_val=str(row[col1]).strip()
    #if 1 we take it as a yes
    if(raw_val=='1')or (raw_val=='1.0'):
        val='yes'
    #else we take it as a no
    else:
        val='no'
    #we append these strings to differ diff types of inference and yes/no to make it sound more meaningfull
    node_A=val.strip().title()+"(Sought Treatment)"
    node_B=str(row[col2]).strip().title()+"(Interference)"
    #now we simply fill our dictionary for initial responses to setup nodes
    if (node_A not in graph):
        graph[node_A]=[]
    if (node_B not in graph):
        graph[node_B]=[]
    #we setup neighbors of each node
    if (node_B not in graph[node_A]):
         graph[node_A].append(node_B)
    if(node_A not in graph[node_B]):
        graph[node_B].append(node_A)
#we print the graph now
print("Dataset Graph\n")
for node,connected_nodes in graph.items():
    print(f"{node} is connected to -->{connected_nodes}")
#we calculate the total number of nodes and edges in the  graph
print(f"the total number of edges in the graph are {len(graph)}\n")
print(f"the total number of edges in the graph are {sum(len(connections) for connections in graph.values())//2}\n")


Dataset Graph

No(Sought Treatment) is connected to -->['Not Applicable To Me(Interference)', 'Sometimes(Interference)', 'Often(Interference)', 'Rarely(Interference)', 'Never(Interference)']
Not Applicable To Me(Interference) is connected to -->['No(Sought Treatment)', 'Yes(Sought Treatment)']
Yes(Sought Treatment) is connected to -->['Sometimes(Interference)', 'Not Applicable To Me(Interference)', 'Often(Interference)', 'Rarely(Interference)', 'Never(Interference)']
Sometimes(Interference) is connected to -->['Yes(Sought Treatment)', 'No(Sought Treatment)']
Often(Interference) is connected to -->['Yes(Sought Treatment)', 'No(Sought Treatment)']
Rarely(Interference) is connected to -->['Yes(Sought Treatment)', 'No(Sought Treatment)']
Never(Interference) is connected to -->['No(Sought Treatment)', 'Yes(Sought Treatment)']
the total number of edges in the graph are 7

the total number of edges in the graph are 10



                                               Phase 1 Reflection

1. How many rows and columns does your dataset have? What does each row represent?
The dataset consists of 1,433 rows and 63 columns. Each row represents a single individual, specifically a tech industry worker, and contains their specific responses to a comprehensive mental health survey.

2. Are there missing values? If yes, which columns have them and how might this affect your analysis?
Yes, there are missing values, particularly in secondary questions like the work interference columns where respondents may have skipped the question. Missing data can negatively affect the analysis because the AI might have to drop those rows entirely or guess the missing information, which can skew the accuracy of our final models.

3. What is the target variable and how many unique classes does it have?
Our target variable is the column asking, "Have you been diagnosed with a mental health condition by a medical professional?". It is a binary classification target, meaning it has exactly two unique classes: "Yes" and "No".

4. Describe the graph you built. What do the nodes and edges represent in the context of your domain?
I built an undirected bipartite graph mapping "Sought Treatment" against "Work Interference." In this domain context, the Nodes represent the unique multiple-choice answers to those questions (e.g., "Yes (Sought Treatment)" or "Often (Interference)"). The Edges represent actual human connections; an edge is only drawn between two nodes if a single employee in the dataset selected both of those specific answers.

PHASE 2: Intelligent Agent and Search Algorithms

Step 1: Define the Agent and State Space


In [34]:
print(graph.keys())

dict_keys(['no(Sought Treatment)', 'Not Applicable To Me(Interference)', 'yes(Sought Treatment)', 'Sometimes(Interference)', 'Often(Interference)', 'Rarely(Interference)', 'Never(Interference)'])


In [35]:
#we define our agent first
class Agent:
    def __init__(self,graph,goal):
        self.graph=graph
        self.goal=goal
    def perceive(self,state):
        # Looks up the current state in the dictionary to find its neighbors
        return self.graph[state]
    def action(self,act):
        # In this simple graph, the "action" is just the name of the next node
        return act
    def goal_test(self,state):
        # Returns True if we reached the goal, otherwise False
        if(state==self.goal):
            return True
    def get_cost(self,state1,state2):
        # As per the manual, every move costs exactly 1
        return 1


Neighbors of yes: ['Sometimes(Interference)', 'Not Applicable To Me(Interference)', 'Often(Interference)', 'Rarely(Interference)', 'Never(Interference)']
Is yes the goal?: None
Is Often the goal?: True
